### PR TITLE
Make generated case classes final by default.

### PR DESCRIPTION
**Pull request by stewSquared**

### CODE REVIEW

File: slick-codegen/src/main/scala/slick/codegen/AbstractSourceCodeGenerator.scala [hunk 1]
```diff
@@ -78,6 +78,7 @@ abstract class AbstractSourceCodeGenerator(model: m.Model)
         ).mkString(", ")
         if(classEnabled){
           val prns = (parents.take(1).map(" extends "+_) ++ parents.drop(1).map(" with "+_)).mkString("")
+          if(caseClassFinal) "final " else "" +
           s"""case class $name($args)$prns"""
```
Comment (cvogt) [inline slick-codegen/src/main/scala/slick/codegen/AbstractSourceCodeGenerator.scala]: that doesn't seem correct. Mind adding some parenthesis?

```
scala> {
     |   if(true) "s" else "c" +
     |   "x"
     | }
res1: String = s
```

Comment (stewSquared) [inline slick-codegen/src/main/scala/slick/codegen/AbstractSourceCodeGenerator.scala]: @cvogt Good catch! I've corrected it.

I suppose the code generator isn't automatically tested anywhere?

Comment (stewSquared) [inline slick-codegen/src/main/scala/slick/codegen/AbstractSourceCodeGenerator.scala]: Ahh, I found them!

**Comment by hgoldwire** [comment]
With slick 3.2.0-RC1 and scala 2.12.1, my codegen-created final case classes fail to compile with "The outer reference in this type test cannot be checked at run time."  I think this is ==> https://issues.scala-lang.org/browse/SI-4440

Changing them to sealed satisfies the compiler, as does dropping final.  I'd be happy with the second option except I can't figure out where/how to set caseClassFinal = false when overriding SourceCodeGenerator.  Any ideas?

Thanks!

- Henry

**Comment by cvogt** [comment]
@hgoldwire see this part of the docs http://slick.lightbend.com/doc/3.1.1/code-generation.html#customization See how we override `classEnabled`? Override it in the same place. (The code generator uses nested classes, which is a good fit for the problem structure.)

**Comment by stewSquared** [comment]
Ahh, right. I didn't catch this because my own codegen is customized. Ideally these case classes should't be in the trait anyway; the type of the row shouldn't be path dependent.

**Comment by stewSquared** [comment]
But until case classes are gen'rated by default in a separate package or object, they shouldn't be final by default; the default code generator shouldn't need overriding.

**Comment by stewSquared** [comment]
@cvogt It looks like we should either replace `final` with `sealed` and leave default to `true` or replace the boolean flag with a String for modifiers. I can write the PR. What do you think?

**Comment by cvogt** [comment]
People can also just `override def code = "final " ++ super.code`. Maybe we don't need this at all?

**Comment by stewSquared** [comment]
Well, it should probably check `classEnabled` in case `super.code` is a type alias.

Admittedly, most people won't require this. I only changed it to make WartRemover happy. 

At a minimum, we should revert; but it's up to you if you'd prefer an update to `sealed` by default.

**Pull request merged**